docs/adrs/0002-uds-core-functional-layers.md

```diff
@@ -0,0 +1,73 @@
+# ADR: UDS Core Functional Layers
+
+Date: 2024-07-17
+
+## Status
+
+Accepted
+
+## Context
+
+Historically, UDS Core intentionally only published a single official Zarf Package to ensure a standard baseline across all UDS environments. We learned from the complex buffet menu-style of Big Bang that it is better to form opinions and test them than to allow unlimited configuration possibilities. We have also found the monorepo structure of UDS Core to be be much simpler for developing UDS Core. However, through user feedback, we have seen cases where more optionality is warranted, but having to modify UDS CLI to accommodate disabling portions of UDS Core is not ideal. We also have a need for an edge-compatible version of UDS Core that is more lightweight and does not include all the components of the full UDS Core package.
+
+Today, we publish both the official UDS Core package and a UDS Slim Dev package, in addition to dev and demo UDS Bundles. The UDS Slim Dev package is a stripped-down version of UDS Core that includes limited services meant for testing UDS Packages. We have also learned that some teams outside of UDS need similar functionality, but slightly different parts of UDS Core, such as keycloak + authsvc.
+
+## Decision
+
+We will replace the existing 2-package (one dev and one prod) structure with a layered set of packages within the existing monorepo. This will allow users to choose the components they need for their specific use case, while maintaining a single source of truth for all UDS Core components. We will not break these out by applications as Big Bang did, but instead by layers of functionality for related/integrated components. This allows more flexibility for which functionality is provided by UDS Core and which is provided by other external packages.
+
+The functional layers will be:
+
+1. UDS Core Base (Istio + Pepr w/UDS Operator & UDS Policies)
+2. UDS Core Identity and Authorization (Keycloak + AuthService)
+3. UDS Core Metrics (Metrics Server)
+4. UDS Core Monitoring (Prometheus + Grafana)
+5. UDS Core Logging (Promtail + Loki)
+6. UDS Core Runtime Security (NeuVector)
+7. UDS Core Backup and Restore (Velero)
+8. UDS Core (combination of all layers)
+
+These layers can be combined as needed, with UDS Core Base serving as the foundation for all deployments. The UDS Core package will continue to offer the complete suite of tools. The only change in the UDS Core repo will be the addition of new package definitions for each functional layer. Versioning will be managed at the monorepo level, with each layer being versioned in lockstep with the rest of UDS Core.
```

Comment:
Given the optionality here how would the handle the user experience for folks deploying a Package that might contain an `sso:` block but not have Identity and Authorization deployed?

Reply:
I think the short answer is that Pepr/UDS Operator would have the knowledge of the cluster state and handle/ignore the relevant sections. So in the case you mentioned the `sso` block would just be ignored _ I don't think we want to shift the complexity to the end user/bundle creator to have to toggle on/off all the pieces.

The challenge here becomes the app configuration. We would also need to find a way to dynamically handle things like [this](https://github.com/defenseunicorns/uds-core/blob/main/src/grafana/values/values.yaml#L23-L25) where the app has to have SSO toggled on/off.

Reply:
I'm think that all sounds technically feasible but trying to consider users who might not know what layers are installed on a cluster.  How do they know that when they deploy their Package certain pieces would be ignored?

Reply:
I think it should at least partially be pretty transparent based on what they picked to deploy 😅 - you deployed without monitoring in your bundle, it shouldn't be overly surprising that you didn't get monitoring on your other components. But definitely documentation around what each piece of the stack includes/provides will be important, and figuring out good ways to represent the UDS config/deployment state. Maybe this is also a good spot where CLI coupling/integration with Core could warn the end user that their chosen stack is missing identity/ask for confirmation this was intentional.

Definitely want to balance this against the expectation that some users will provide their own alternatives for pieces of core or intentionally opt out of pieces (i.e. they want a lightweight version). We don't want the information/warnings to be aggressive or confusing to users that intentionally are skipping pieces.

Reply:
Some kind of warning might ease my mind but in general I don't think we should be assuming folks who are authoring Package CRs are always the same people who built/deployed the UDS Core bundle.

Reply:
Yeah I wasn't thinking they are the same persona - in my mind the `Package` CR spec would stay the same the whole time, the only thing changing is the operator having some logic be conditional. So the bundle creator doesn't necessarily know/care about what was in the `Package` CR - they only know that they did or didn't include monitoring/sso in their bundle and the effects of that.

Reply:
To pull the bottom line out of this thread - I think the nontrivial problem here is we need to figure out how to dynamically modify upstream helm charts based on the "uds cluster state". Two things discussed synchronously (neither of which feel great):
1. Pepr could mutate all the things we need to dynamically configure - this could be any number of resources, and would have to be specific per resource/package (i.e. a specific mutate for Grafana's configuration to add/remove the keycloak pieces). The downside of this is that each specific mutate must be baked in - which isn't a good solve for SWF, etc, on top of core.
2. UDS-CLI could handle this _somehow_ - maybe this looks like dynamic configuration/referencing of which values to use based on a cluster state resource/lookup. This might mean baking business logic into CLI for what capabilities necessitate a change in configuration (or at least having individual packages produce metadata that could be used by CLI)?

Would be curious to hear @jeff-mccoy or other's thoughts here.

Reply:
A key point we need to make going forward, and maybe adjust this doc to state that more plainly, it is assumed you will deploy UDS Core with UDS CLI as a bundle in the end. Whether that is only the complete uds-core package or otherwise. The idea being here at bundle create time you already know what you have/what you need. My thinking on those that required SSO--let's just keep that as a required assumption for now. Start with it in the docs, later on we can add a field of some sort that says "hey my guy, this package needs this this package"...not saying full-on DAG, but one-layer deep for these functional layers would not be hard. So that if someone tried to create a bundle without uds-core-base, for example, it would yell at them.

Reply:
Updated ADR to reflect docs for dependencies + future CLI changes to codify dependencies. For initial implementation we can document SSO as a requirement for any core packages that use it.